### PR TITLE
[pickers] Keep the calendar header and content in sync when switching locale

### LINKED ISSUE #12963: [pickers] Changing locale between a monday-first or sunday-first language breaks indicated order of weekdays
### Steps to reproduce

Link to live example: https://stackblitz.com/edit/react-xqjmtx?file=Demo.tsx

Steps:
1. See which day of the week 1st is. (e.g. in May 2024, it's on Wednesday)
2. Switch the language to "PT", on top.
3. Notice how 1st is now on a "Tuesday". Every day was shifted by 1, which completely messes up the calendar.

### Current behavior

After the executing the steps, the calendar displays the month days in the wrong week days.
Sounds like it's related to the fact that Portuguese has a Monday-first week and English has a Sunday-first week.

### Expected behavior

Days should not be inaccurate after switching languages.
**The same issue can be reproduced with DateCalendar, for example.**

### Context

_No response_

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>
```
  System:
    OS: Windows 10 10.0.19045
  Binaries:
    Node: 18.16.0 - D:\nodejs\node.EXE
    npm: 9.8.0 - D:\nodejs\npm.CMD    
    pnpm: Not Found
  Browsers:
    Chrome: Not Found
    Edge: Chromium (123.0.2420.97)    
  npmPackages:
    @emotion/react: ^11.11.4 => 11.11.4
    @emotion/styled: ^11.11.5 => 11.11.5
    @mui/base:  5.0.0-beta.40
    @mui/core-downloads-tracker:  5.15.15
    @mui/material: ^5.15.15 => 5.15.15
    @mui/private-theming:  5.15.14
    @mui/styled-engine:  5.15.14
    @mui/system:  5.15.15
    @mui/types:  7.2.14
    @mui/utils:  5.15.14
    @mui/x-date-pickers: ^7.3.1 => 7.3.1
    @types/react: ^18.2.15 => 18.2.22
    react: ^18.2.0 => 18.2.0
    react-dom: ^18.2.0 => 18.2.0
    styled-components:  5.3.11
    typescript: ^5.0.2 => 5.2.2
```
</details>

**Search keywords**: locale, static-date-picker

### PR DESCRIPTION
**Pull request by flaviendelangle**
Fixes #12963

In `AdapterDateFns` we have the following:

```ts
  public startOfWeek = (value: Date): Date => {
    return startOfWeek(value, { locale: this.locale });
  };
```

But on the adapters where the locale is stored on the date object, we didn't set the locale of the adapter to the value passed.
So if the locale of the value and the value of the adapter were different, it was the locale of the value that was being used.

### CODE REVIEW

File: packages/x-date-pickers/src/DateCalendar/tests/DateCalendar.spec.tsx [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import moment, { Moment } from 'moment';
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/DateCalendar/tests/DateCalendar.spec.tsx:2]: Not related to this PR, I just noticed that we had 2 spec files for this component

File: packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx [hunk 2]
```diff
@@ -574,7 +574,7 @@ export function DayCalendar<TDate extends PickerValidDate>(inProps: DayCalendarP
             key={i.toString()}
             variant="caption"
             role="columnheader"
-            aria-label={utils.format(utils.addDays(startOfCurrentWeek, i), 'weekday')}
+            aria-label={utils.format(weekday, 'weekday')}
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx:577]: No idea why we were re-doing the day shift here, but I spent some time finding out why my `aria-label` were not updated in the test whereas the UI was good in the doc :laughing:

File: packages/x-date-pickers/src/internals/utils/date-utils.ts [hunk 3]
```diff
@@ -203,6 +203,7 @@ export const getWeekdays = <TDate extends PickerValidDate>(
   utils: MuiPickersAdapter<TDate>,
   date: TDate,
 ) => {
-  const start = utils.startOfWeek(date);
+  const cleanDate = utils.setLocaleToValue(date);
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/internals/utils/date-utils.ts:206]: The alternative solution would be to do `utils.` inside `utils.startOfWeek` and remove it's usage in `utils.getWeekArray` and here.

Advantages:
- No new public method
- No impact on Date Fns adapter which does not care about this

Disavantages:
- Could this be needed for other methods than `utils.startOfWeek`? In which case setting the locale explicitly before processing the value is probably simpler than doing it in tons of adapter methods
- Small overhead on `utils.startOfWeek`

**Review by arthurbalduini — APPROVED** [review]
👏

File: packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx (hunk 2 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx:576]: Nice optimization. 👍

**Review by LukasTy — COMMENTED** [review]
Nice fix! 👏 
Sorry for a late review. 🙈

**Pull request merged**